src/Temporalio/Worker/WorkflowInstance.cs

```diff
@@ -541,29 +557,59 @@ public WorkflowActivationCompletion Activate(WorkflowActivation act)
-                        var sortedJobs = act.Jobs.OrderBy(j =>
+
+                        // We only sort jobs in legacy event loop logic, modern relies on Core
+                        List<WorkflowActivationJob> jobs;
+                        if (applyModernEventLoopLogic)
                         {
-                            switch (j.VariantCase)
+                            jobs = act.Jobs.ToList();
+                        }
+                        else
+                        {
+                            jobs = act.Jobs.OrderBy(j =>
                             {
-                                case WorkflowActivationJob.VariantOneofCase.NotifyHasPatch:
-                                case WorkflowActivationJob.VariantOneofCase.UpdateRandomSeed:
-                                    return 1;
-                                case WorkflowActivationJob.VariantOneofCase.SignalWorkflow:
-                                case WorkflowActivationJob.VariantOneofCase.DoUpdate:
-                                    return 2;
-                                case WorkflowActivationJob.VariantOneofCase.InitializeWorkflow:
-                                    return 3;
-                                default:
-                                    return 4;
-                            }
-                        }).ToList();
-                        // We can trust jobs are deterministically ordered by core
-                        foreach (var job in sortedJobs)
+                                switch (j.VariantCase)
+                                {
+                                    case WorkflowActivationJob.VariantOneofCase.NotifyHasPatch:
+                                    case WorkflowActivationJob.VariantOneofCase.UpdateRandomSeed:
+                                        return 1;
+                                    case WorkflowActivationJob.VariantOneofCase.SignalWorkflow:
+                                    case WorkflowActivationJob.VariantOneofCase.DoUpdate:
+                                        return 2;
+                                    case WorkflowActivationJob.VariantOneofCase.InitializeWorkflow:
+                                        return 3;
+                                    default:
+                                        return 4;
+                                }
+                            }).ToList();
+                        }
+
+                        // Apply each job
+                        foreach (var job in jobs)
                         {
                             Apply(job);
-                            // Run scheduler once. Do not check conditions when patching or querying.
-                            var checkConditions = job.NotifyHasPatch == null && job.QueryWorkflow == null;
+                            // We only run the scheduler after each job in legacy event loop logic
+                            if (!applyModernEventLoopLogic)
```

Comment:
Presumably, this can't change within an activation right? If so, can we bring this outside the loop over the jobs instead of checking for each job

Reply:
The legacy behavior is `RunOnce` inside-loop (so we have to preserve it here) and the new behavior is `RunOnce` after-loop (which is seen below). So we have to leave the inside-loop here for older workflows.

Reply:
Oh I missed the `Apply(job);` line above, at the beginning of the loop 👍